batch2/batch/driver/main.py

```diff
@@ -0,0 +1,265 @@
+import asyncio
+import concurrent
+import logging
+import os
+import traceback
+
+from aiohttp import web
+import kubernetes as kube
+
+import prometheus_client as pc
+from prometheus_async.aio.web import server_stats
+
+from hailtop.auth import async_get_userinfo
+from hailtop.config import get_deploy_config
+
+# import uvloop
+
+from ..batch import Batch, Job
+from ..log_store import LogStore
+from ..batch_configuration import KUBERNETES_TIMEOUT_IN_SECONDS, REFRESH_INTERVAL_IN_SECONDS, \
+    POD_VOLUME_SIZE, INSTANCE_ID, BATCH_IMAGE, BATCH_NAMESPACE
+from ..database import BatchDatabase
+
+from .driver import Driver
+from .k8s import K8s
+
+# uvloop.install()
+
+log = logging.getLogger('batch')
+
+POD_EVICTIONS = pc.Counter('batch_pod_evictions', 'Count of batch pod evictions')
+
+log.info(f'KUBERNETES_TIMEOUT_IN_SECONDS {KUBERNETES_TIMEOUT_IN_SECONDS}')
+log.info(f'REFRESH_INTERVAL_IN_SECONDS {REFRESH_INTERVAL_IN_SECONDS}')
+log.info(f'POD_VOLUME_SIZE {POD_VOLUME_SIZE}')
+log.info(f'INSTANCE_ID = {INSTANCE_ID}')
+log.info(f'BATCH_IMAGE = {BATCH_IMAGE}')
+
+routes = web.RouteTableDef()
+
+deploy_config = get_deploy_config()
+
+
+@routes.get('/healthcheck')
+async def get_healthcheck(request):  # pylint: disable=W0613
+    return web.Response()
+
+
+@routes.patch('/api/v1alpha/batches/{user}/{batch_id}/close')
+async def close_batch(request):
+    user = request.match_info['user']
+    batch_id = int(request.match_info['batch_id'])
+    batch = await Batch.from_db(request.app, batch_id, user)
+    if not batch:
+        raise web.HTTPNotFound()
+    asyncio.ensure_future(batch._close_jobs())
+    return web.Response()
+
+
+@routes.patch('/api/v1alpha/batches/{user}/{batch_id}/cancel')
+async def cancel_batch(request):
+    user = request.match_info['user']
+    batch_id = int(request.match_info['batch_id'])
+    batch = await Batch.from_db(request.app, batch_id, user)
+    if not batch:
+        raise web.HTTPNotFound()
+    asyncio.ensure_future(batch._cancel_jobs())
+    return web.Response()
+
+
+@routes.delete('/api/v1alpha/batches/{user}/{batch_id}')
+async def delete_batch(request):
+    user = request.match_info['user']
+    batch_id = int(request.match_info['batch_id'])
+    batch = await Batch.from_db(request.app, batch_id, user)
+    if not batch:
+        raise web.HTTPNotFound()
+    # would prefer for this to be in front end, but then Batch.from_db
+    # won't be able to find batch
```

Comment:
Yes, but my general model is that batch2 pushes information into the database, which batch2-driver consumes and executes.  Splitting an action across two services has to be thought through carefully because it becomes concurrent and subject to partial failure.  So I was hoping to put the durable state change in batch2 and if the batch2 endpoint fails, at some point one of the heal loops will clean things up.

I will still probably make this change in a later PR.

Reply:
Oh I see now.